llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -19512,8 +19515,13 @@ RISCVTargetLowering::shouldExpandAtomicRMWInIR(AtomicRMWInst *AI) const {
     return AtomicExpansionKind::None;
 
   unsigned Size = AI->getType()->getPrimitiveSizeInBits();
-  if (Size == 8 || Size == 16)
-    return AtomicExpansionKind::MaskedIntrinsic;
+  if (Size == 8 || Size == 16) {
+    if (!Subtarget.hasStdExtZabha())
+      return AtomicExpansionKind::MaskedIntrinsic;
+    if (AI->getOperation() == AtomicRMWInst::Nand)
```

Comment:
I find the Nand conditions confusing as written.

How about the following:
```
  [...]
  unsigned Size = AI->getType()->getPrimitiveSizeInBits();
  if (AI->getOperation() == AtomicRMWInst::Nand && Subtarget.hasStdExtZacas() &&
      (Size >= 32 || Subtarget.hasStdExtZabha()))
    return AtomicExpansionKind::CmpXChg;
  if (Size < 32 && !Subtarget.hasStdExtZabha())
    return AtomicExpansionKind::MaskedIntrinsic;
  return AtomicExpansionKind::None;
```

Reply:
I think that causes i8/i16 Nand to return None with Zbha and no Zacas. We need to keep MaskIntrin.

How about 

```
  unsigned Size = AI->getType()->getPrimitiveSizeInBits();                       
  if (AI->getOperation() == AtomicRMWInst::Nand) {                               
    if (Subtarget.hasStdExtZacas() &&                                            
        (Size >= 32 || Subtarget.hasStdExtZabha()))                              
      return AtomicExpansionKind::CmpXChg;                                       
    return AtomicExpansionKind::MaskedIntrinsic;                                 
  }                                                                              
                                                                                 
  if (Size < 32 && !Subtarget.hasStdExtZabha())                                  
    return AtomicExpansionKind::MaskedIntrinsic;                                 
                                                                                 
  return AtomicExpansionKind::None;
```

Reply:
That's still wrong :(